livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,423 @@
+
+import asyncio
+import dataclasses
+import json
+import os
+import time
+from contextlib import suppress
+from dataclasses import dataclass
+from typing import List, Optional, Union, Literal
+from urllib.parse import urlencode
+
+import aiohttp
+from livekit import rtc
+from livekit.agents import stt, utils
+from livekit.agents.utils import AudioBuffer
+
+
+ENGLISH = "en"
+
+# Define bytes per frame for different encoding types
+bytes_per_frame = {
+    "pcm_s16le": 2,
+    "pcm_mulaw": 1,
+}
+
+
+@dataclass
+class STTOptions:
+    sample_rate: Optional[int] = None
+    word_boost: Optional[List[str]] = None
+    encoding: Optional[Literal["pcm_s16le", "pcm_mulaw"]] = None
+    disable_partial_transcripts: bool = False
+    enable_extra_session_information: bool = False
+    end_utterance_silence_threshold: Optional[int] = None
+    buffer_size_seconds: Optional[float] = None
+    token_expires_in: Optional[int] = None
+
+
+class STT(stt.STT):
+    def __init__(
+        self,
+        *,
+        api_key: Optional[str] = None,
+        sample_rate: Optional[int] = 16000,
+        word_boost: Optional[List[str]] = None,
+        encoding: Optional[str] = "pcm_s16le",
+        disable_partial_transcripts: bool = False,
+        enable_extra_session_information: bool = False,
+        end_utterance_silence_threshold: int = 1000,
+        http_session: Optional[aiohttp.ClientSession] = None,
+        token_expires_in: int = 3600,
+        buffer_size_seconds: float = 0.2,
+    ):
+        super().__init__(streaming_supported=True)
+        api_key = api_key or os.environ.get("ASSEMBLYAI_API_KEY")
+        if api_key is None:
+            raise ValueError("AssemblyAI API key is required. " \
+                             "Pass one in via the `api_key` parameter, " \
+                             "or set it as the `ASSEMBLYAI_API_KEY` environment variable")
+        self._api_key = api_key
+
+        self._opts = STTOptions(
+            sample_rate=sample_rate,
+            word_boost=word_boost,
+            encoding=encoding,
+            disable_partial_transcripts=disable_partial_transcripts,
+            enable_extra_session_information=enable_extra_session_information,
+            buffer_size_seconds=buffer_size_seconds,
+            token_expires_in=token_expires_in,
+            end_utterance_silence_threshold=end_utterance_silence_threshold,
+        )
+        self._session = http_session
+        self._token_expires_in = token_expires_in
+
+    @property
+    def session(self) -> aiohttp.ClientSession:
+        if not self._session:
+            self._session = utils.http_session()
+        return self._session
+
+    async def recognize(
+        self,
+        *,
+        buffer: AudioBuffer,
+    ) -> stt.SpeechEvent:
+        raise NotImplementedError("Not implemented")
+
+    def stream(
+        self,
+        *,
+        language: Optional[str] = None,
+    ) -> "SpeechStream":
+        config = dataclasses.replace(self._opts)
```

Comment:
Is a copy of `self._opts` (rather than directly passing) used just to prevent future bugs?